### PR TITLE
Sanitizers

### PR DESCRIPTION
**Pull request by r1viollet**
**What does this PR do?**:

Sanitizers
- Add a configuration for sanitizers
- Run TSan, UBsan, ASan
- Ensure unit tests run with aligned configurations (and remove the CMake files)

Various TSan fixes
- Adjust atomic operations to clarify some of the sequencing

Elf parser fixes
- Fix to ignore some out of bound elf symbols
- Addition of a test for the elf parser

CI
- Daily report of sanitizer issues

**Motivation**:

Better static analysis

**Additional Notes**:

You can now run various build configurations through
```
./gradlew testasan
./gradlew testubsan
./gradlew testdebug
./gradlew testrelease
```

Additionally the library can be used to run on other types of workloads.
```
UBSAN_OPTIONS="halt_on_error=1 abort_on_error=1 print_stacktrace=1" LSAN_OPTIONS="detect_leaks=0" ASAN_OPTIONS="detect_stack_use_after_return=1" LD_PRELOAD=/usr/lib/gcc/x86_64-linux-gnu/11/libasan.so DD_SERVICE=test_r1 java -javaagent:./temp/dd-java-agent.jar -Ddd.profiling.enabled=true -Ddd.profiling.ddprof.enabled=true -Ddd.profiling.upload.period=10 -Ddd.profiling.start-force-first=true -Ddd.profiling.ddprof.debug.lib=/home/r1viollet/dd/java-profiler/ddprof-lib/build/lib/main/asan/linux/x64/libjavaProfiler.so -jar ./temp/renaissance-gpl-0.15.0.jar akka-uct -r 5
```

To avoid deadly signal issues on ubuntu 22, I had to run 
```
sudo sysctl vm.mmap_rnd_bits=28
```
because of https://github.com/actions/runner-images/issues/9491

**How to test the change?**:

This is run within our CI daily.

**For Datadog employees**:
- [ ] If this PR touches code that signs or publishes builds or packages, or handles
  credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [X] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: ddprof-lib/src/main/cpp/stackFrame_x64.cpp [hunk 1]
```diff
@@ -95,7 +96,14 @@ bool StackFrame::unwindStub(instruction_t* entry, const char* name, uintptr_t& p
         pc = ((uintptr_t*)sp)[0] - 1;
         sp += 8;
         return true;
-    } else if (entry != NULL && *(unsigned int*)entry == 0xec8b4855) {
+    }
+    if(entry == nullptr) {
+        return false;
+    }
+    if (reinterpret_cast<uintptr_t>(entry) % alignof(unsigned int) != 0) {
```
Comment (r1viollet) [inline ddprof-lib/src/main/cpp/stackFrame_x64.cpp:103]: I am not 100% sure this is true. To be verified.

File: ddprof-lib/src/test/run_tests.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+TMPDIR="${TMPDIR:-$(dirname $(mktemp))}"
+
+function build_test() {
+  NAME=$1
+  echo "Building test ${NAME}..."
+  # shellcheck disable=SC2038
+  files=$(find ${PROJECT_DIR}/build/obj/main/${CONFIG_NAME} -name "*.o" -print | tr '\n' ' ')
+  libdir=${GTEST_DIR}/configs/${CONFIG_NAME}/lib
+  g++ -std=c++11 -o ${TMPDIR}/${NAME}_${CONFIG_NAME} ${PROJECT_DIR}/src/test/cpp/${NAME}.cpp \
```
Comment (r1viollet) [inline ddprof-lib/src/test/run_tests.sh:22]: I would not go this route ;-)

File: gradle/configurations.gradle [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+  }
+  return null
+}
+
+def locateLibasan() {
+  return locateLibrary('libasan')
+}
+
+def locateLibtsan() {
+  def libtsan_preinit = locate('libtsan_preinit.o')
```
Comment (r1viollet) [inline gradle/configurations.gradle:61]: I do not require this to build locally. Is this required in CI ?

File: ddprof-lib/src/main/cpp/profiler.cpp [hunk 4]
```diff
@@ -1143,7 +1143,7 @@ Error Profiler::start(Arguments& args, bool reset) {
 
         _state = RUNNING;
         _start_time = time(NULL);
-        _epoch++;
+        __atomic_add_fetch(&_epoch, 1, __ATOMIC_SEQ_CST);
```
Comment (r1viollet) [inline ddprof-lib/src/main/cpp/profiler.cpp:1146]: I am not sure how important it is to put a SEQ_CST here. What is the use of epoch ?

Comment (richardstartin) [reply ddprof-lib/src/main/cpp/profiler.cpp:1146]: it's only used to check the recording hasn't rolled since a wallclock sample was taken

Comment (r1viollet) [reply ddprof-lib/src/main/cpp/profiler.cpp:1146]: Ok, that means if we do not retrieve the correct value, we are fine with it. I will go back to no ordering constraints:
 `__ATOMIC_RELAXED`

File: ddprof-lib/src/main/cpp/callTraceStorage.cpp [hunk 5]
```diff
@@ -50,7 +50,9 @@ class LongHashTable {
         if (table != NULL) {
             table->_prev = prev;
             table->_capacity = capacity;
-            table->_size = 0;
+            // The reset is not useful with the anon mmap setting the memory is zeroed.
+            // However this silences a false positive and should not have a performance impact.
+            table->clear();
```
Comment (r1viollet) [inline ddprof-lib/src/main/cpp/callTraceStorage.cpp:55]: I could make this a conditional change. I do not think it matters in terms of performance, as it is a memset at startup and on growth.

File: ddprof-lib/src/main/cpp/stackFrame_x64.cpp (hunk 1 above)
Comment (r1viollet) [reply ddprof-lib/src/main/cpp/stackFrame_x64.cpp:103]: I think this is a reasonable assumption

File: .github/workflows/nightly.yml [hunk 6]
```diff
@@ -1,11 +1,33 @@
...
+        with:
+          name: failures
+          path: ./artifacts
+      - name: Report failures
+        run: |
+          find ./artifacts -name 'failures_*' -exec cat {} \; > ./artifacts/failures.txt
+          scenarios=$(cat ./artifacts/failures.txt | tr '\n' ',')
+          echo "Failed scenarios: $scenarios"
+
+#          curl -X POST "${{ secrets.SLACK_WEBHOOK }}" \
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply .github/workflows/nightly.yml:30]: Yes. Once we are ready to merge we should uncomment this block

File: ddprof-lib/src/main/cpp/callTraceStorage.cpp (hunk 5 above)
Comment (jbachorik) [reply ddprof-lib/src/main/cpp/callTraceStorage.cpp:55]: Considering that the next sub-table is 2x size of the previous one the number of memset calls should not get high.

**Review by jbachorik — APPROVED** [review]
LGTM

**Pull request merged**